### ISSUE TITLE
Can't download `setuptools`

### ISSUE THREAD

**Issue by MinchinWeb** [pypa/virtualenv#807]
`virtualenv` fails to create the virtual environment as it seems unable to find `setuptools` to download an install in.

Python 3.5.0, system install, WIndows 10, virtualenv ~~1.3.2~~ 13.1.2

`virtualenv` works as expected on the same machine using Python 3.4.3.

This is the output:
```
> virtualenv env5
Using base prefix 'c:\\program files\\python 3.5'
New python executable in env5\Scripts\python.exe
Installing setuptools, pip, wheel...
  Complete output from command C:\Users\User\Docume...5\Scripts\python.exe -c "import sys, pip; sys...d\"] + sys.argv[1:]))" setuptools pip wheel:
  Ignoring indexes: https://pypi.python.org/simple
Collecting setuptools
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  The repository located at None is not a trusted or secure host and is being ignored. If this repository is available via HTTPS it is recommended to use HTTPS instead, otherwise you may silence this warning and allow it anyways with '--trusted-host None'.
  Could not find a version that satisfies the requirement setuptools (from versions: )
No matching distribution found for setuptools
----------------------------------------
...Installing setuptools, pip, wheel...done.
Traceback (most recent call last):
  File "c:\program files\python 3.5\lib\runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\program files\python 3.5\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Program Files\Python 3.5\Scripts\virtualenv.exe\__main__.py", line 9, in <module>
  File "c:\program files\python 3.5\lib\site-packages\virtualenv.py", line 832, in main
    symlink=options.symlink)
  File "c:\program files\python 3.5\lib\site-packages\virtualenv.py", line 1004, in create_environment
    install_wheel(to_install, py_executable, search_dirs)
  File "c:\program files\python 3.5\lib\site-packages\virtualenv.py", line 969, in install_wheel
    'PIP_NO_INDEX': '1'
  File "c:\program files\python 3.5\lib\site-packages\virtualenv.py", line 910, in call_subprocess
    % (cmd_desc, proc.returncode))
OSError: Command C:\Users\User\Docume...5\Scripts\python.exe -c "import sys, pip; sys...d\"] + sys.argv[1:]))" setuptools pip wheel failed with error code 1
```

**Comment by Ivoz** [comment]
Could you confirm the virtualenv version that you used? The latest is `13.1.2`, but the one you specified, `1.3.2`, was released in December 2008.

**Comment by MinchinWeb** [comment]
I guess it was *really* late...

I'm using `virtualenv` 13.1.2 and have also reproduced the issue with 13.1.1.

**Comment by efagerberg** [comment]
Hello this is also happening to myself as well.

**Comment by MinchinWeb** [comment]
@efagerberg I think you're right. If I create a virtual environment with Python 3.4, activate it, and then try to run `virtualenv` from there (so the Python execacutable is at `C:\tmp\some d\env4\Scripts\Python.exe`), it fails with the same issue.

**Comment by MinchinWeb** [comment]
If it's helpful, here are the results of `virtualenv`'s tests. I cloned the current version (cba89a145a36a836b8060752e47dee89fd079ad4), and navigated to the root directory.
```
S:\Documents\GitHub\virtualenv\>py.test
============================= test session starts =============================
platform win32 -- Python 3.5.0, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: S:\Documents\GitHub\virtualenv, inifile: 
collected 9 items

test.txt s
tests\test_virtualenv.py .F.F....

================================== FAILURES ===================================
_________________ test_resolve_interpreter_with_absolute_path _________________

mock_exists = <MagicMock name='exists' id='414201680728'>

    @patch('os.path.exists')
    def test_resolve_interpreter_with_absolute_path(mock_exists):
        """Should return absolute path if given and exists"""
        mock_exists.return_value = True
        virtualenv.is_executable = Mock(return_value=True)
    
        exe = virtualenv.resolve_interpreter("/usr/bin/python42")
    
>       assert exe == "/usr/bin/python42", "Absolute path should return as is"
E       AssertionError: Absolute path should return as is
E       assert 'C:/usr/bin/python42' == '/usr/bin/python42'
E         - C:/usr/bin/python42
E         ? --
E         + /usr/bin/python42

tests\test_virtualenv.py:26: AssertionError
______________ test_resolve_interpreter_with_invalid_interpreter ______________

mock_exists = <MagicMock name='exists' id='414201836432'>

    @patch('os.path.exists')
    def test_resolve_interpreter_with_invalid_interpreter(mock_exists):
        """Should exit when with absolute path if not exists"""
        mock_exists.return_value = True
        virtualenv.is_executable = Mock(return_value=False)
    
        try:
            virtualenv.resolve_interpreter("/usr/bin/python42")
            assert False, "Should raise exception"
        except SystemExit:
            pass
    
>       mock_exists.assert_called_with("/usr/bin/python42")

tests\test_virtualenv.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
c:\program files\python 3.5\lib\site-packages\mock\mock.py:937: in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

value = AssertionError("Expected call: exists('/usr/bin/python42')\nActual call: exists('C:/usr/bin/python42')",)
from_value = None

>   ???
E   AssertionError: Expected call: exists('/usr/bin/python42')
E   Actual call: exists('C:/usr/bin/python42')

<string>:2: AssertionError
---------------------------- Captured stdout call -----------------------------
The executable C:/usr/bin/python42 (from --python=C:/usr/bin/python42) is not executable
=============== 2 failed, 6 passed, 1 skipped in 14.35 seconds ================
```

**Comment by frp** [comment]
For me, removing spaces fixes the problem. So the actual bug is in the code which processes paths.

**Comment by YSunLIN** [comment]
I found the bugs in the function install_wheel
```python
def install_wheel(project_names, py_executable, search_dirs=None):
    if search_dirs is None:
        search_dirs = file_search_dirs()

    wheels = find_wheels(['setuptools', 'pip'], search_dirs)
    pythonpath = os.pathsep.join(wheels)
    findlinks = ':'.join(search_dirs)

    cmd = [
        py_executable, '-c',
        'import sys, pip; sys.exit(pip.main(["install", "--ignore-installed"] + sys.argv[1:]))'
    ] + project_names
    logger.start_progress('Installing %s...' % (', '.join(project_names)))
    logger.indent += 2
    try:
        call_subprocess(cmd, show_stdout=False,
            extra_env = {
                'PYTHONPATH': pythonpath,
                'JYTHONPATH': pythonpath,  # for Jython < 3.x
                'PIP_FIND_LINKS': findlinks,
                'PIP_USE_WHEEL': '1',
                'PIP_PRE': '1',
                'PIP_NO_INDEX': '1'
            }
        )
    finally:
        logger.indent -= 2
        logger.end_progress()
```

```python
findlinks = ':'.join(search_dirs)
```
Here is the problem, **search_dirs** could include space, but they join **search_dirs** by space, OMG.
And the ultimate problem is that pip env varible PIP_FIND_LINKS doesn't support the space in windows path.
When you fix this bug, I think they should modify pip and virtualenv code in the same time.OMG.

**Comment by MinchinWeb** [comment]
On Windows, when paths with a space are entered at the commandline, they have to be surrounded by quotes. I.e.
```
C:> cd "c:\some path with spaces\more"
```
Does `PIP_FIND_LINKS` work the same way? So pass it a string like:
```
C:\path1 C:\another_path C:\a\longer\path "C:\path with a space\more" C:\path2
```

**Comment by YSunLIN** [comment]
I have tried, and failed again.So I ```PIP_FIND_LINKS``` doesn't work in the same way actually.

**Comment by xflr6** [comment]
Same problem here with Python 2.7 installed under `C:\Program Files\Python27\`:

```
C:\Users\User\Desktop>python -i -m virtualenv venv
(...)
OSError: Command C:\Users\User\Desktop\venv\Scripts\python.exe -c "import sys, p
ip; sys...d\"] + sys.argv[1:]))" setuptools pip wheel failed with error code 1
>>> import pdb; pdb.pm()
> c:\program files\python27\lib\site-packages\virtualenv.py(910)call_subprocess()
-> % (cmd_desc, proc.returncode))
(Pdb) cmd
['C:\\Users\\User\\Desktop\\venv\\Scripts\\python.exe', '-c', 'import sys, pip;
sys.exit(pip.main(["install", "--ignore-installed"] + sys.argv[1:]))', 'setuptoo
ls', 'pip', 'wheel']
(Pdb) q
>>>
```

Workaround using Windows short path names (`SPAM~1`) in `PIP_FIND_LINKS`:

```
C:\Users\User\Desktop>virtualenv --no-setuptools venv
New python executable in venv\Scripts\python.exe

C:\Users\User\Desktop>set PIP_FIND_LINKS=c:\\PROGRA~1\\Python27\\Lib\\site-packages\\virtualenv_support
C:\Users\User\Desktop>set PIP_USE_WHEEL=1
C:\Users\User\Desktop>set PIP_NO_INDEX=1
C:\Users\User\Desktop>set PIP_PRE=1
C:\Users\User\Desktop>set PYTHONPATH=C:\\Program Files\\Python27\\lib\\site-packages\\virtualenv_support\\setuptools-18.2-py2.py3-none-any.whl;C:\\Program Files\\Python27\\lib\\site-packages\\virtualenv_support\\pip-7.1.2-py2.py3-none-any.whl

C:\Users\User\Desktop>venv\Scripts\python.exe -c "import sys, pip; sys.exit(pip.main(['install', '--ignore-installed', 'setuptools', 'pip', 'wheel']))"
Ignoring indexes: https://pypi.python.org/simple
Collecting setuptools
Collecting pip
Collecting wheel
Installing collected packages: setuptools, pip, wheel
Successfully installed pip-7.1.2 setuptools-18.2 wheel-0.24.0

C:\Users\User\Desktop>venv\Scripts\activate.bat
(venv) C:\Users\User\Desktop>pip list
pip (7.1.2)
setuptools (18.2)
wheel (0.24.0)
```

So this approach would make it fixable without modifying `pip`.

There is already some [short path using code](https://github.com/pypa/virtualenv/blob/e657022e5a138a8827d212e8e2f6910416a63986/virtualenv.py#L893-L906) in `virtualenv`. Maybe [this](https://stackoverflow.com/questions/23598289/how-to-get-windows-short-file-name-in-python/23598461#23598461) can help to improve the details.

**Comment by dstufft** [comment]
should be fixed now with #830
